### ISSUE TITLE
Proposal: Add new `stop` hook return value

### ISSUE THREAD

**Issue by larshesel** [vernemq/vernemq#955]
Currently a decision hook can return `ok`, `{ok, Modifiers}`, `next` and `{error, Reason}`. With MQTT 5 we sometimes would like to terminate a plugin chain and disconnect the client but in a normal way. For instance, in MQTT 5.0 it is perfectly legal to tell send `reason_code = use_another_server | server_moved, server_reference=server_id` in a CONNACK or DISCONNECT packet.

Such use cases don't express themselves well using the `{ok, Modifiers}` or `{error, Reason}` return values as these are used respectively for the success case where the client stays connected and the error case where something 'abnormal' happened and the client must be disconnected. Currently the closest option is to embed the use case above into the `{error, Reason}` return value, but this seems semantically wrong as it's not really an error, and therefore it's potentially a source of confusion and could make plugins hard(er) to understand.

So the proposal is to add a `{stop, Modifiers}` (though, we should find some other name than modifiers) to support uses cases where no error happened, but we'd like to gracefully terminate the session and optionally send something back to the client.

Let me know what you think.

**Comment by dergraf** [comment]
As far as I understand such a return tuple would only apply to the M5 hooks. Therefore we have plenty of freedom to design them properly... it could be something like `{stop, Reason::atom(), Properties::map()}` or similar.

**Comment by larshesel** [comment]
Exactly, the current use-case is for the MQTT 5.0 hooks, but I could imagine such a return tuple could be handy in MQTT 3.1.1 plugins as well, as a 'normal' stop would imply that the client is disconnected without logging any errors.

Your proposal `{stop, Reason::atom(), Properties::map()}` is interesting, I would change the last part to something more general like `{stop, Reason::atom(), Args|Opts|..::map()}` and not pass properties directly (like we don't do this in other return tuples, but specify explicitly what one can return - many things corresponding directly to properties). This three tuple is perhaps even a generalization of the `error` tuple as the `Reason` could be for instance `normal` or some error.

In any case, since you're not directly opposed to the idea I'll make a spike where I try to incorporate a `stop` tuple into for example `auth_on_register_m5` and `auth_on_publish_m5` to handle the `server_moved` and `use_another_server` use case mentioned above.

**Comment by codeadict** [comment]
I like this idea, currently, I use `{error, Reason}` for some plugin that needs to disconnect the client. Using error kind of makes the code confusing for others and this `stop` makes it perfect.  Using it in the MQTT 3.1.1 hooks will be beneficial as well and leaves less unicorns between how plugins are written for different MQTT versions. `{stop, Reason::atom(), Options::map()}` sounds good to me.

**Comment by dergraf** [comment]
This all sounds good to me too!
